text/3470-crabi-v1.md

```diff
@@ -0,0 +1,661 @@
+
+crABI supports slices (references to arrays) as long as crABI supports the
+element type.
+
+Slices translate to a by-value struct containing two fields: a pointer to the
+element type, and a `size_t` number of elements, in that order.
+
+For instance:
+
+```rust
+extern "crabi" fn func(buf: &mut [u8]);
+```
+
+is equivalent to:
+
+```c
+struct u8_slice {
+    uint8_t *data;
+    size_t len;
+};
+extern void func(struct u8_slice buf);
+```
+
+## `&str` - borrowed counted UTF-8 string
+
+crABI supports `&str`, a borrowed counted UTF-8 string. This type must always
+contain valid UTF-8. (Use a different type for non-UTF-8 strings.) These are
+not typically NUL-terminated.
+
+crABI handles this type equivalently to a byte slice `&[u8]`.
+
+## `Box<[T]>` - boxed slice
+
+crABI supports passing ownership of a slice (an array with size determined at
+runtime). To represent this via the C ABI, crABI passes this equivalently to a
+C struct containing an element pointer and a `size_t` number of elements, in
+that order.
+
+For instance:
+
+```rust
+extern "crabi" fn func(data: Box<[u32]>) -> Box<[f64]>;
+```
+
+is equivalent to:
+
+```c
+struct u32_boxed_slice {
+    uint32_t *data;
+    size_t len;
+};
+struct f64_boxed_slice {
+    double *data;
+    size_t len;
+};
+extern struct f64_boxed_slice func(struct u32_boxed_slice data);
+```
+
+## `Box<str>` - boxed counted UTF-8 string
+
+crABI supports passing an owned string. To represent this via the C ABI, crABI
+passes this equivalently to a boxed slice of type `Box<[u8]>`.
+
+## `[T; N]` - fixed-size array by value
+
+crABI supports passing a fixed-size array by value (as opposed to by
+reference). To represent this via the C ABI, crABI treats this equivalently to
+a C array of the same type passed by value.
+
+Note that this means C code can use an array directly in a context where it
+will be interpreted by-value (such as in a struct field), but needs to use a
+structure with a single array field in contexts where it would otherwise be
+interpreted as a pointer (such as in a function argument or return value).
+
+For instance:
+
+```rust
+extern "crabi" fn func(rgb: [u16; 3])
+```
+
+is equivalent to:
+
+```c
+struct func_rgb_arg {
+    uint16_t array[3];
+};
+extern void func(struct func_rgb_arg rgb);
+```
+
+Note that crABI does *not* pass the length, since it's a compile-time constant;
+the recipient must also know the correct size. (Use one of the slice-based
+types for a type with a runtime-determined length.)
+
+## `&[T; N]` - fixed-size array by reference
+
+crABI supports passing a fixed-size array by reference. crABI represents this
+as a pointer to the element type.
+
+C can represent this as an array (e.g. `uint16_t rgb[3]`) in contexts where
+that already implies passing by pointer (such as a function argument or return
+value), or translate it explicitly to a pointer (e.g `uint16_t (*rgb)[3]`) in
+contexts where just writing the array would imply by-value (such as a struct
+field).
```

Comment:
The C types `uint16_t *` and `uint16_t (*)[3]` are distinct types. The wording here clearly specifies that the former be used, yet an example uses the latter.

It should be made clear which one to use. And I believe `&[T; N]` should always be lowered to `T (*)[N]`.

> C can represent this as an array (e.g. `uint16_t rgb[3]`) in contexts where
> that already implies passing by pointer (such as a function argument or return
> value)

This can simply be removed. It is true that function parameters declared as arrays are pointers and that returning arrays is illegal, but that doesn’t concern crABI.